File: src/main/java/com/chattriggers/ctjs/minecraft/wrappers/objects/inventory/Item.java
package com.chattriggers.ctjs.minecraft.wrappers.objects.inventory;

import com.chattriggers.ctjs.minecraft.wrappers.Client;
import com.chattriggers.ctjs.minecraft.wrappers.Player;
import com.chattriggers.ctjs.minecraft.wrappers.objects.block.Block;
import com.chattriggers.ctjs.minecraft.wrappers.objects.Entity;
import lombok.Getter;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.item.ItemStack;
import org.lwjgl.opengl.GL11;

//#if MC<=10809
//$$ import net.minecraft.client.renderer.entity.RenderItem;
//#else
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.client.util.ITooltipFlag;
import com.chattriggers.ctjs.minecraft.wrappers.World;
//#endif

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Item {
    @Getter private net.minecraft.item.Item item;
    @Getter private ItemStack itemStack;

    /* Constructors */
    /**
     * Creates an Item object from a Minecraft ItemStack input.
     *
     * @param itemStack the Minecraft ItemStack
     */
    public Item(ItemStack itemStack) {
        ItemStack newItemStack = itemStack;
        if (itemStack == null)
            newItemStack = new ItemStack(new Block(0).getBlock());

        this.item = newItemStack.getItem();
        this.itemStack = newItemStack;
    }

    /**
     * Creates an Item object from a string name input.
     *
     * @param itemName the name of the item
     */
    public Item(String itemName) {
        this.item = net.minecraft.item.Item.getByNameOrId(itemName);
        this.itemStack = new ItemStack(this.item);
    }

    /**
     * Creates an Item object from an integer ID input, 0 for air.
     *
     * @param itemID the ID of the item
     */
    public Item(int itemID) {
        this.item = net.minecraft.item.Item.getItemById(itemID);
        this.itemStack = new ItemStack(this.item);
    }

    /**
     * Creates an Item object from a {@link Block} object input.
     *
     * @param block the {@link Block}
     */
    public Item(Block block) {
        this.item = net.minecraft.item.Item.getItemFromBlock(block.getBlock());
        this.itemStack = new ItemStack(this.item);
    }

    /**
     * Creates an Item object from an EntityItem.
     *
     * @param entityItem the EntityItem.
     */
    public Item(EntityItem entityItem) {
        //#if MC<=10809
        //$$ this.item = entityItem.getEntityItem().getItem();
        //$$ this.itemStack = entityItem.getEntityItem();
        //#else
        this.item = entityItem.getItem().getItem();
        this.itemStack = entityItem.getItem();
        //#endif
    }

    /**
     * Created an Item object from an Entity.
     * Has to be wrapping an EntityItem.
     *
     * @param entity the Entity
     */
    public Item(Entity entity) {
        if (entity.getEntity() instanceof EntityItem) {
            //#if MC<=10809
            //$$ this.item = ((EntityItem) entity.getEntity()).getEntityItem().getItem();
            //$$ this.itemStack = ((EntityItem) entity.getEntity()).getEntityItem();
            //#else
            this.item = ((EntityItem) entity.getEntity()).getItem().getItem();
            this.itemStack = ((EntityItem) entity.getEntity()).getItem();
            //#endif
        } else {
            throw new IllegalArgumentException("Entity is not of type EntityItem");
        }
    }
    /* End of constructors */

    /**
     * Gets the ID of the item.
     *
     * @return the ID
     */
    public int getID() {
        return net.minecraft.item.Item.getIdFromItem(this.item);
    }

    /**
     * Sets the stack size of the item.
     *
     * @param stackSize the stack size
     * @return the Item
     */
    public Item setStackSize(int stackSize) {
        this.itemStack = new ItemStack(this.item, stackSize);
        return this;
    }

    /**
     * Gets the stack size of the item.
     *
     * @return the stack size
     */
    public int getStackSize() {
        //#if MC<=10809
        //$$ return this.itemStack.stackSize;
        //#else
        return this.itemStack.getCount();
        //#endif
    }

    /**
     * Gets the item's unlocalized name.<br>
     * Example: <code>tile.wood</code>
     *
     * @return the item's unlocalized name
     */
    public String getUnlocalizedName() {
        return this.item.getUnlocalizedName();
    }

    /**
     * Gets the item's registry name.<br>
     * Example: <code>minecraft:planks</code>
     *
     * @return the item's registry name
     */
    public String getRegistryName() {
        //#if MC<=10809
        //$$ return this.item.getRegistryName();
        //#else
        return this.item.getRegistryName().toString();
        //#endif
    }

    /**
     * Gets the item's stack display name.<br>
     * Example: <code>Oak Wood Planks</code>
     *
     * @return the item's stack display name
     */
    public String getName() {
        return getID() == 0 ? "air" : this.itemStack.getDisplayName();
    }

    /**
     * Returns a map of the enchantment name to level for all
     * the enchantments on an item.
     *
     * @return the map of enchantments
     */
    public Map<String, Integer> getEnchantments() {
        Map rawEnchants = EnchantmentHelper.getEnchantments(this.itemStack);
        Map<String, Integer> mappedEnchants = new HashMap<>();

        for (Object enchantObj : rawEnchants.entrySet()) {
            Map.Entry<Integer, Integer> rawEnchant = (Map.Entry<Integer, Integer>) enchantObj;

            //#if MC<=10809
            //$$ Enchantment enchant = Enchantment.getEnchantmentById(rawEnchant.getKey());
            //#else
            Enchantment enchant = Enchantment.getEnchantmentByID(rawEnchant.getKey());
            //#endif

            mappedEnchants.put(enchant.getName().replace("enchantment.", ""), rawEnchant.getValue());
        }

        return mappedEnchants;
    }

    /**
     * Gets if the item can be enchanted.
     *
     * @return true if the item can be enchanted
     */
    public Boolean isEnchantable() {
        return this.itemStack.isItemEnchantable();
    }

    /**
     * Gets if the item is enchanted.
     *
     * @return true if the item is enchanted
     */
    public Boolean isEnchanted() {
        return this.itemStack.isItemEnchanted();
    }

    /**
     * Gets the json string of the item's lore.
     *
     * @return the json string of the lore
     */
    public String getItemNBT() {
        return this.itemStack.serializeNBT().toString();
    }

    /**
     * Gets the metadata of the item.
     *
     * @return the metadata
     */
    public int getMetadata() {
        return this.itemStack.getMetadata();
    }

    /**
     * Gets if the item can be placed on a {@link Block}.
     *
     * @param block the {@link Block} to place the item on
     * @return true if the item can be placed on the {@link Block}
     */
    public Boolean canPlaceOn(Block block) {
        return this.itemStack.canPlaceOn(block.getBlock());
    }

    /**
     * Gets if the item can harvest a {@link Block}.
     *
     * @param block the {@link Block} for the item to harvest
     * @return true if the item can harvest the {@link Block}
     */
    public Boolean canHarvest(Block block) {
        //#if MC<=10809
        //$$ return this.itemStack.canHarvestBlock(block.getBlock());
        //#else
        return this.itemStack.canHarvestBlock(
                World.getWorld().getBlockState(block.getBlockPos())
        );
        //#endif
    }

    /**
     * Gets if the item can destroy a {@link Block}.
     *
     * @param block the {@link Block} for the item to destroy
     * @return true if the item can destroy the {@link Block}
     */
    public Boolean canDestroy(Block block) {
        return this.itemStack.canDestroy(block.getBlock());
    }

    /**
     * Gets the items durability, i.e. the number of uses left
     *
     * @return the items durability
     */
    public int getDurability() {
        return this.getMaxDamage() - this.getDamage();
    }

    /**
     * Gets the item's damage.
     *
     * @return the damage value
     */
    public int getDamage() {
        return this.itemStack.getItemDamage();
    }

    /**
     * Sets the item's damage.
     *
     * @param damage the damage value
     */
    public Item setDamage(int damage) {
        this.itemStack.setItemDamage(damage);

        return this;
    }

    /**
     * Gets the items max damage.
     *
     * @return the items max damage
     */
    public int getMaxDamage() {
        return this.itemStack.getMaxDamage();
    }

    /**
     * Checks if the item can take damage.
     *
     * @return true if the item can take damage
     */
    public boolean isDamagable() {
        return this.itemStack.isItemStackDamageable();
    }

    /**
     * Gets the item lore/tooltip information in a list.
     *
     * @return the item lore
     */
    public List<String> getLore() {
        //#if MC<=10809
        //$$ return this.itemStack.getTooltip(Player.getPlayer(), Client.getMinecraft().gameSettings.advancedItemTooltips);
        //#else
        return this.itemStack.getTooltip(Player.getPlayer(), ITooltipFlag.TooltipFlags.ADVANCED);
        //#endif
    }

    /**
     * Checks whether another Item is the same as this one.
     * It compares id, stack size, and durability.
     *
     * @param obj the object to compare to
     * @return whether the objects are equal
     */
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Item) {
            Item other = ((Item) obj);

            return getID() == other.getID()
                    && getStackSize() == other.getStackSize()
                    && getDamage() == other.getDamage();
        }

        return false;
    }

    @Override
    public String toString() {
        return this.itemStack.toString();
    }

    /**
     * Renders the item icon to the client's overlay.
     *
     * @param x the x location
     * @param y the y location
     * @param scale the scale
     */
    public void draw(float x, float y, float scale) {
        float finalX = x / scale;
        float finalY = y / scale;

        RenderItem itemRenderer = Client.getMinecraft().getRenderItem();

        GlStateManager.pushMatrix();

        GlStateManager.scale(scale, scale, 1f);
        GlStateManager.translate(finalX, finalY, 0);
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);

        RenderHelper.enableStandardItemLighting();
        RenderHelper.enableGUIStandardItemLighting();
        itemRenderer.zLevel = 200.0F;

        itemRenderer.renderItemIntoGUI(getItemStack(), 0, 0);

        GlStateManager.popMatrix();
    }

    /**
     * Renders the item icon to the client's overlay.
     *
     * @param x the x location
     * @param y the y location
     */
    public void draw(float x, float y) {
        draw(x, y, 1);
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/wrappers/objects/inventory/Item.kt
package com.chattriggers.ctjs.minecraft.wrappers.objects.inventory

import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer
import com.chattriggers.ctjs.minecraft.objects.message.TextComponent
import com.chattriggers.ctjs.minecraft.wrappers.Client
import com.chattriggers.ctjs.minecraft.wrappers.Player
import com.chattriggers.ctjs.minecraft.wrappers.World
import com.chattriggers.ctjs.minecraft.wrappers.objects.entity.Entity
import com.chattriggers.ctjs.minecraft.wrappers.objects.block.BlockType
import com.chattriggers.ctjs.minecraft.wrappers.objects.inventory.nbt.NBTTagCompound
import com.chattriggers.ctjs.utils.kotlin.External
import com.chattriggers.ctjs.utils.kotlin.MCBlock
import com.chattriggers.ctjs.utils.kotlin.MCBlockPos
import com.chattriggers.ctjs.utils.kotlin.toIdentifier
import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import net.fabricmc.fabric.api.renderer.v1.RendererAccess
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.pattern.CachedBlockPosition
import net.minecraft.client.item.TooltipContext
import net.minecraft.enchantment.Enchantment
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.ItemEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.Registry
import net.minecraft.world.WorldView
import net.minecraft.item.Item as MCItem

@External
class Item {
    val item: MCItem
    var itemStack: ItemStack

    constructor(itemStack: ItemStack) {
        item = itemStack.item
        this.itemStack = itemStack
    }

    constructor(itemName: String) {
        item = Registry.ITEM[itemName.toIdentifier()]
        itemStack = ItemStack(item)
    }

    constructor(itemID: Int) {
        item = MCItem.byRawId(itemID)
        itemStack = ItemStack(item)
    }

    constructor(block: BlockType) {
        item = block.mcBlock.asItem()
        itemStack = ItemStack(item)
    }

    // TODO("fabric")
    // constructor(entityItem: ItemEntity) {
    //     item = entityItem.entityItem.item
    //     itemStack = entityItem.entityItem
    // }

    // TODO("fabric")
    // /**
    //  * Create an Item object from an Entity.
    //  * Has to be wrapping an EntityItem.
    //  *
    //  * @param entity the Entity
    //  */
    // constructor(entity: Entity) {
    //     if (entity.entity is EntityItem) {
    //         //#if MC<=10809
    //         item = entity.entity.entityItem.item
    //         itemStack = entity.entity.entityItem
    //         //#else
    //         //$$ item = entity.entity.item.item
    //         //$$ itemStack = entity.entity.item
    //         //#endif
    //     } else {
    //         throw IllegalArgumentException("Entity is not of type EntityItem")
    //     }
    // }

    // TODO(VERIFY)
    fun getTextComponent() = TextComponent(itemStack.name)

    fun getRawNBT() = itemStack.orCreateNbt.toString()

    fun getNBT() = NBTTagCompound(itemStack.orCreateNbt)

    // TODO(BREAKING): remove this
    // @Deprecated("Use the better-named method", ReplaceWith("getNBT"))
    // fun getItemNBT(): NBTTagCompound = getNBT()

    fun getID(): Int = MCItem.getRawId(item)

    fun setStackSize(stackSize: Int) = apply {
        itemStack = ItemStack(item, stackSize)
    }

    fun getStackSize(): Int = itemStack.maxCount

    /**
     * Gets the item's unlocalized name.
     * Example: tile.wood
     *
     * @return the item's unlocalized name
     */
    // TODO(VERIFY)
    fun getUnlocalizedName(): String = item.name.toString()

    /**
     * Gets the item's registry name.
     * Example: minecraft:planks
     *
     * @return the item's registry name
     */
    fun getRegistryName(): String = Registry.ITEM.getId(item).toString()

    /**
     * Gets the item's stack display name.
     * Example: Oak Wood Planks
     *
     * @return the item's stack display name
     */
    // TODO(VERIFY)
    fun getName(): String = if (getID() == 0) "air" else item.getName(itemStack).toString()

    // TODO(BREAKING): Keys now have namespaces
    fun getEnchantments(): Map<String, Int> {
        return EnchantmentHelper.get(itemStack).mapKeys {
            Registry.ENCHANTMENT.getId(it.key).toString()
        }
    }

    fun isEnchantable(): Boolean = itemStack.isEnchantable

    fun isEnchanted(): Boolean = itemStack.hasEnchantments()

    // TODO("fabric")
    // fun getMetadata(): Int = itemStack.

    // TODO("fabric"): Verify this actually does what I want in the methods below
    class CachedBlockType(view: WorldView, val type: BlockType) : CachedBlockPosition(view, BlockPos.ORIGIN, false) {
        override fun getBlockState(): BlockState {
            return type.getDefaultState()
        }
    }

    fun canPlaceOn(block: BlockType): Boolean {
        val world = World.getWorld() ?: return false
        return itemStack.canPlaceOn(world.tagManager, CachedBlockType(world, block))
    }

    // TODO(BREAKING): Remove this method
    // fun canHarvest(block: BlockType): Boolean {
    //     return itemStack.canHarvestBlock(block.mcBlock)
    // }

    fun canDestroy(block: BlockType): Boolean {
        val world = World.getWorld() ?: return false
        return itemStack.canDestroy(world.tagManager, CachedBlockType(world, block))
    }

    /**
     * Gets the item's durability, i.e. the number of uses left
     *
     * @return the item's durability
     */
    fun getDurability(): Int = getMaxDamage() - getDamage()

    fun getDamage(): Int = itemStack.damage

    fun setDamage(damage: Int) = apply {
        itemStack.damage = damage
    }

    fun getMaxDamage(): Int = itemStack.maxDamage

    fun isDamagable(): Boolean = itemStack.isDamageable

    fun getLore(): List<String> {
        val context = if (Client.getMinecraft().options.advancedItemTooltips) {
            TooltipContext.Default.ADVANCED
        } else TooltipContext.Default.NORMAL

        return itemStack.getTooltip(Player.getPlayer(), context).map { TextComponent(it).getFormattedText() }
    }

    // TODO("fabric")
    // /**
    //  * Renders the item icon to the client's overlay.
    //  *
    //  * @param x the x location
    //  * @param y the y location
    //  * @param scale the scale
    //  */
    // @JvmOverloads
    // fun draw(x: Float = 0f, y: Float = 0f, scale: Float = 1f) {
    //     val itemRenderer = Client.getMinecraft().renderItem
    //
    //     GlStateManager.scale(scale, scale, 1f)
    //     GlStateManager.translate(x / scale, y / scale, 0f)
    //     GlStateManager.color(1f, 1f, 1f, 1f)
    //
    //     RenderHelper.enableStandardItemLighting()
    //     RenderHelper.enableGUIStandardItemLighting()
    //
    //     itemRenderer.zLevel = 200f
    //     itemRenderer.renderItemIntoGUI(itemStack, 0, 0)
    //
    //     Renderer.finishDraw()
    // }

    /**
     * Checks whether another Item is the same as this one.
     * It compares id, stack size, and durability.
     *
     * @param other the object to compare to
     * @return whether the objects are equal
     */
    override fun equals(other: Any?): Boolean {
        return other is Item &&
                getID() == other.getID() &&
                getStackSize() == other.getStackSize() &&
                getDamage() == other.getDamage()
    }

    override fun hashCode(): Int {
        var result = item.hashCode()
        result = 31 * result + itemStack.hashCode()
        return result
    }

    override fun toString(): String = itemStack.toString()
}
